lab1/README.md

```diff
@@ -152,4 +152,380 @@ func main() {
+        }
+    }
+}
+```
+As above statement, you can stop multiple worker nodes with a single context simultaneously. You can also implement a graceful shutdown to cancel the running jobs through this approach.
+
+Of course, you can also declare multiple contexts and **`cancel`** functions, waiting for goroutines to complete their jobs with **`cancel`** and **`Done`**:
+```go
+func main() {
+	ctx1, cancel1 := context.WithCancel(context.Background())
+	ctx2, cancel2 := context.WithCancel(context.Background())
+
+	go func() {
+		task1()
+		cancel1() //call cancel1 function
+	}()
+
+	go func() {
+		task2()
+		cancel2()
+	}()
+	
+	<-ctx1.Done() //blocked until the context is canceled
+	<-ctx2.Done()
+	
+	//-----------------
+	// keep going down to execute
+}
+
+func task1() {
+	fmt.Println("Starting job1")
+	time.Sleep(3 * time.Second)
+	fmt.Println("Finished job1")
+}
+
+func task2() {
+	fmt.Println("Starting job2")
+	time.Sleep(2 * time.Second)
+	fmt.Println("Finished job2")
+}
+```
+
+## Exercise
+- What is atomic operation?
+- Will the following work in parallel code?
+```go
+var a int64
+
+func main() {
+    var wg sync.WaitGroup
+    wg.Add(1)
+    
+    go func() {
+        increment()
+        wg.Done()
+    }()
+
+    wg.Wait()
+
+    fmt.Println("Final value of a:", a)
+}
+
+func increment() {
+    for i := 0; i < 100000000; i++ {
+        atomic.AddInt64(&a, 1)
+    }
+}
+```
+- What do the **`wg.Add(1)`** and **`wg.Done()`** do in the above statement? And what does the **`1`** repersent?
+- Please define a Counter struct with an integer field and a sync.Mutex, then implement a function to increment the counter safely.
+- Why is there a fetal error in following code?
+```go=
+func main() {
+    var intChan chan int
+    fmt.Println(intChan)
+    intChan <- 10
+}
+```
+- In Topic: **Select**-**Setup Timeout mechanism**, We offered a sample code that showed how to setup **`timeout`** with **`select`**. Actually, there is a potentail error (Hint: memory leak) because of **`time.After`** usage. Please describe the reason for the error and how to fix it.
+
+## Answer
+- What is atomic operation?
+An atomic operation is a low-level operation that is completed in a single step relative to other threads or processes. Without locks, there are only simple CPU instructions that cannot be further subdivided into smaller steps.
+- Will the following work in parallel code?
+```go
+var a int64
+
+func main() {
+    var wg sync.WaitGroup
+    wg.Add(1)
+    
+    go func() {
+        increment()
+        wg.Done()
+    }()
+
+    wg.Wait()
+
+    fmt.Println("Final value of a:", a)
+}
+
+func increment() {
+    for i := 0; i < 100000000; i++ {
+        atomic.AddInt64(&a, 1)
+    }
+}
+```
+Yes, The **`atomic.AddInt64`** ensures that the increment operation on the **`a`** is performed atomically. This means that the increment operation cannot be interrupted, ensuring the correctness and expected result even when it's accessed by multiple goroutines.
```

Comment:
In this example, the use of atomic operation is redundant.
It's because of there is only one goroutine accessed to the variable a.

The example below would be better:
```
// bad code
var a int64

func main() {
    var wg sync.WaitGroup
    wg.Add(1)

    go func() {
        increment()
        wg.Done()
    }()

    wg.Wait()

    fmt.Println("Final value of a:", a)
}

func increment() {
    for i := 0; i < 100000000; i++ {
        a = a + 1
    }
}
```
```
// good code
var a int64

func main() {
    var wg sync.WaitGroup
    wg.Add(1)

    go func() {
        increment()
        wg.Done()
    }()

    wg.Wait()

    fmt.Println("Final value of a:", a)
}

func increment() {
    for i := 0; i < 100000000; i++ {
        go atomic.AddInt64(&a, 1)
    }
}
```

Reply:
Thank you for advice